drivers/can/can_stm32h7.c

```diff
@@ -0,0 +1,333 @@
+};
+
+struct can_stm32h7_data {
+	struct can_mcan_data mcan_data;
+};
+
+int can_stm32h7_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	int rate_tmp;
+
+	rate_tmp = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE);
+
+	if (rate_tmp == LL_RCC_PERIPH_FREQUENCY_NO) {
+		LOG_ERR("Can't read core clock");
+		return -EIO;
+	}
+
+	*rate = rate_tmp;
+
+	LOG_DBG("rate=%d", *rate);
+
+	return 0;
+}
+
+int can_stm32h7_get_max_filters(const struct device *dev, enum can_ide id_type)
+{
+	if (id_type == CAN_STANDARD_IDENTIFIER) {
+		return NUM_STD_FILTER_DATA;
+	} else {
+		return NUM_EXT_FILTER_DATA;
+	}
+}
+
+int can_stm32h7_clock_enable(const struct device *dev)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	LL_RCC_SetFDCANClockSource(LL_RCC_FDCAN_CLKSOURCE_PLL1Q);
+
+	if (clock_control_on(clk,
+		(clock_control_subsys_t *) &cfg->pclken) != 0) {
+		LOG_ERR("failure enabling clock");
+		return -EIO;
+	}
+
+	if (!LL_RCC_PLL1Q_IsEnabled()) {
+		LOG_ERR("PLL1Q clock must be enabled!");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+void can_stm32h7_set_state_change_cb(const struct device *dev,
+				     can_state_change_callback_t cb,
+				     void *user_data)
+{
+	struct can_stm32h7_data *data = dev->data;
+
+	data->mcan_data.state_change_cb = cb;
+	data->mcan_data.state_change_cb_data = user_data;
+}
+
+static int can_stm32h7_init(const struct device *dev)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_stm32h7_data *data = dev->data;
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = cfg->msg_sram;
+	int ret;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("CAN pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	ret = can_stm32h7_clock_enable(dev);
+	if (ret) {
+		return ret;
+	}
+
+	ret = can_mcan_init(dev, mcan_cfg, msg_ram, mcan_data);
+	if (ret) {
+		return ret;
+	}
+
+	cfg->config_irq();
+
+	return ret;
+}
+
+enum can_state can_stm32h7_get_state(const struct device *dev,
+				     struct can_bus_err_cnt *err_cnt)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+
+	return can_mcan_get_state(mcan_cfg, err_cnt);
+}
+
+int can_stm32h7_send(const struct device *dev, const struct zcan_frame *frame,
+		     k_timeout_t timeout, can_tx_callback_t callback,
+		     void *user_data)
+{
+	const struct can_stm32h7_config *cfg = dev->config;
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_stm32h7_data *data = dev->data;
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = cfg->msg_sram;
+
+	return can_mcan_send(mcan_cfg, mcan_data, msg_ram, frame, timeout,
+			     callback, user_data);
```

Comment:
suggestion: I think this reads better:

```suggestion
	const struct can_stm32h7_config *cfg = dev->config;
	struct can_stm32h7_data *data = dev->data;

	return can_mcan_send(&cfg->mcan_cfg, &data->mcan_data, cfg->msg_sram,
			     frame, timeout, callback, user_data);
```

Reply:
Agreed, adopted everywhere.